### DISCUSSION TITLE
TCSC in pandapower

### DISCUSSION THREAD

**Discussion by lschewe** [e2nIEE/pandapower#1288] [Ideas]
We want to model a power grid that contains a TCSC (something like [this](https://library.e.abb.com/public/fdf0b019e1fe08a48325771f002dbfc5/A02-0158.pdf)). We do not see how to model this in PP and would be happy for any advice.

**Comment by jkisse** [comment]
This element should also be an active part in OPF calculations.

**Comment by lthurner** [comment]
I think for power flow this could be modeled with a controller that controls an impedance element.

However, you mentioned that this should be an OPF element. Can you describe how this element should behave, i.e. what would be the optimization variable and constraints?

**Comment by friederikemeier** [comment]
I agree, this could be done with a controller for normal power flow calculation.
Do you need this to be considered in the OPF as well? If so, do you want it to be optimized? What variables of it should be optimized? Can the shunt element be used for that? Should it be modeled as a PQ or a PV node?
What works fine in small scale networks is to iterate controller loops and opf calculation. Imagine you wrote your own local controller TCSCctrl which is serving as a "local" controller. And you could also have a "central" controller conducting an OPF. 
Then you could setup the simulation like this:

```
CentralController(net, ..., level=[0,2])
TCSCctrl(net, ..., level=[1, 2])
run_control(net)

```
And then think about a senseful structure of order & level. We sometimes do it like this for transformer control, as it can not be respected in the OPF directly.

**Comment by rbolgaryn** [comment]
Dear @lschewe ,

have you tried implementing the functionality of the device in a controller? Has it been a viable option for you?

Roman
